Make as many services private as possible

nicolas-grekas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | not yet
| Fixed tickets | #23822
| License       | MIT
| Doc PR        | -

- [x] review the list of currently public services and validate which one should be turned private (noted as such with the `container.private` tag in this PR)
- [x] rebase on top of #24103
- [x] implement the behavior described in https://github.com/symfony/symfony/issues/23822#issuecomment-327208485
- [x] add tests

List of services that will be kept public:

### Remaining public aliases

cache.app_clearer <- for the cache:pool:clear command to clear app pools
$commandId <- for non-lazy commands which are get() at runtime
router <- for the base controller
templating <- for the base controller

### Remaining public services

test.client <- for WebTestCase
security.password_encoder <- for use in ContainerAware classes
security.authentication_utils <- for use in ContainerAware classes
filesystem <- for use in ContainerAware classes
kernel <- for use in ContainerAware classes
translator <- for use in ContainerAware classes
validator <- for use in ContainerAware classes
cache_clearer <- for the cache:clear command
cache_warmer <- required to bootstrap the kernel
cache.app <- for userland only
cache.global_clearer <- for the cache:pool:clear command to clear all pools
cache.system <- for userland only
data_collector.dump <- required to have dump() work very early when booting the kernel
event_dispatcher <- required to wire console apps
form.factory <- for the base controller
http_kernel <- required to bootstrap the kernel
profiler <- used in tests
request_stack <- for the base controller
routing.loader <- used by routing
security.authorization_checker <- for the base controller
security.csrf.token_manager <- for the base controller
security.token_storage <- for the base controller
serializer <- for the base controller
session <- for the base controller
state_machine.abstract <- userland state machines
workflow.abstract <- userland workflows
twig <- for the base controller
twig.controller.exception <- controllers referenced by routing
twig.controller.preview_error <- controllers referenced by routing
var_dumper.cloner <- required to have dump() work very early when booting the kernel
web_profiler.controller.exception <- controllers referenced by routing
web_profiler.controller.profiler <- controllers referenced by routing
web_profiler.controller.router <- controllers referenced by routing

nicolas-grekas committed:
Make as many services private as possible

chalasr commented on REMAINING.PUBLIC.md:
```diff
@@ -0,0 +1,70 @@
+security.password_encoder
+session.handler
+session.storage
+session.storage.filesystem
```
flagged as "for BC" https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/Resources/config/session.xml#L72..L73
should we deprecate it on 3.4? (can't find the PR)

stof commented on REMAINING.PUBLIC.md:
```diff
@@ -0,0 +1,70 @@
+annotation_reader
+cache.app_clearer
+$commandId
+debug.templating.engine.php
```
the debug one does not need to be public IMO. You should never get it by its `debug.*` id (the `templating.engine.php` should stay public though)

stof commented on REMAINING.PUBLIC.md:
```diff
@@ -0,0 +1,70 @@
+router
+security.access.decision_manager
+security.acl.cache
+security.acl.dbal.connection
```
this should not need to be public (if you need the DBAL connection yourselves dynamically, you should get it using the public ids given in DoctrineBundle)

stof commented on REMAINING.PUBLIC.md:
```diff
@@ -0,0 +1,70 @@
+templating
+templating.loader
+translator
+twig.loader
```
even this one should become private IMO. There is no reason to access it dynamically at runtime IMO.

chalasr commented on REMAINING.PUBLIC.md:
```diff
@@ -0,0 +1,70 @@
+
+annotation_reader
+cache.app_clearer
+$commandId
```
must remain public, this is for non-lazy commands which are `get()` at runtime

stof commented on REMAINING.PUBLIC.md:
```diff
@@ -0,0 +1,70 @@
+cache.system
+cache_warmer <- required to bootstrap the kernel
+data_collector.cache
+data_collector.dump
```
Do collectors actually need to be public ? AFAIK, they are needed only by the profiler, and so could become private

stof commented on REMAINING.PUBLIC.md:
```diff
@@ -0,0 +1,70 @@
+
+### Remaining public services
+
+cache.adapter.* <- pools are private by default when created by config so not a pb
```
I suggest adapters should become private in 4.0

stof commented on REMAINING.PUBLIC.md:
```diff
@@ -0,0 +1,70 @@
+security.csrf.token_manager
+security.firewall
+security.token_storage
+security.validator.user_password
```
should become private, like other constraint validators

stof commented on REMAINING.PUBLIC.md:
```diff
@@ -0,0 +1,70 @@
+serializer
+session
+state_machine.abstract <- state machines created by config are public
+workflow.abstract <- workflows created by config are public
```
can't we make them private instead ? The documented way to get them is to use the registry AFAIK, not to get the service from the container

stof commented on REMAINING.PUBLIC.md:
```diff
@@ -0,0 +1,70 @@
+translator
+twig
+twig.controller.exception
+twig.controller.preview_error
```
these are required to be public due to the routing referencing them, right ?

stof commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/assets.xml:
```diff
@@ -17,23 +18,27 @@
 
-        <service id="assets.path_package" class="Symfony\Component\Asset\PathPackage" abstract="true" public="true">
+        <service id="assets.path_package" class="Symfony\Component\Asset\PathPackage" abstract="true">
+            <tag name="container.private" />
```
the place where we actually need to apply this tag is in child services

nicolas-grekas commented:
@chalasr @stof thanks for the first round of review, comments addressed.

In order to allow deprecating public aliases, I changed the implementation to not rely on a tag anymore. Instead, I added `Alias/Definition::isPrivate()`. The PR is ready for a second round review. It only needs tests now.

ro0NL reviewed (commented):
Awesome. Fixed rather elegant :+1: is setPrivate an internal detail now? should it?

nicolas-grekas commented on REMAINING.PUBLIC.md:
```diff
@@ -0,0 +1,62 @@
+serializer <- for the base controller
+session <- for the base controller
+state_machine.abstract <- state machines created by config are public
+workflow.abstract <- workflows created by config are public
```
@lyrixx @Nyholm should we plan to make these services private in 4.0 and tell ppl that injecting workflow.registry is the way to go?

lyrixx replied:
No. The registry is only here for convenience in twig.
Users should inject the right workflow in other services. Adding an extra layer will make the core harder to understand and to test. It's exactly like doctrine. People should inject the right repository, not the whole Doctrine Manager.

nicolas-grekas replied:
OK, then we can still make them private, isn't it?

stof replied:
@lyrixx if you want them to inject it in the service, it can be private. Public is here for dynamic get.

But if they are meant to be the main services of the workflow component, they should indeed stay public for now.

lyrixx replied:
No, because we need to access it from a controller or a command.
Not everyone are using controller/command as a service.

nicolas-grekas commented on REMAINING.PUBLIC.md:
```diff
@@ -0,0 +1,62 @@
+translation.extractor
+translation.loader
+translation.reader
+translation.writer
```
@Nyholm do you think we should plan making private these translation.*/translator services in 4.0?

stof replied:
`translator` must stay public. Otherwise, people using ContainerAware classes cannot use the translator layer anymore

nicolas-grekas replied:
note added. any comment I could add on the remaining aliases/services?

lyrixx replied:
IMHO, all theses services (above) could be private. except the `security.authentication_utils` (for use in ContainerAware classes)

nicolas-grekas commented:
@ro0NL setPrivate is OK for eg bundles to me so they can do the same as we do here.

ro0NL commented:
Guessed so. But are we sure about `setPublic` & `setPrivate` doing different things, which is here to stay for the long run. This requires documentation, but i think it needs a better name...

nicolas-grekas commented:
> setPublic & setPrivate doing different things

they don't need to do different thing in 4.0, that's the solution.

nicolas-grekas commented:
And in fact, they already don't do different thing - expect for internal details.

ro0NL commented:
Yeah but `setPublic(false)` wont enable `isPrivate() === true`.. right? That's unexpected.

nicolas-grekas commented:
> but setPublic(false) wont enable isPrivate() === true.. right

locally, that's correct - making it so would just discard any benefits of introducing it.
But the `ResolveChildDefinitionsPass` turns isPrivate() is `setPublic(false)` in the end, so semantically, they're the same.
The special behavior related to isPrivate is 3.4 only and tailored to the purpose of making currently public services private in 4.0.
For this reason also, "setPrivate" cannot be controller via a (yaml/xml) loader - only a DI ext can. So only "advanced" use can use it.

stof commented:
I think we need to improve the documentation about the difference between `setPublic(false)` and `setPrivate(true)` in the phpdoc (the phpdoc will be the place where advanced users will learn about it IMO, rather than learning it in symfony-docs, but the current phpdoc does not allow to understand the difference and why they should use one rather than the other)

stof commented on src/Symfony/Component/DependencyInjection/Alias.php:
```diff
@@ -40,10 +41,42 @@ public function isPublic()
+     *
+     * When set, the "private" state has a higher precedence than "public".
+     * You can use it since version 3.4 to flag the "public" visibility
+     * as deprecated and turn the alias private in 4.0.
```
This description looks weird to me, compared to the actual implementation. When reading it, I understand I should use `public=true, private=true` to make the public visibility deprecated.

However, the implementation actually allows to use `public=false, private=true` as well to make a service which stays available through the BC layer of accessing private services (so preventing inlining and removal).
Is this expected ?

nicolas-grekas replied:
Correct, and the implemented behavior is the expected one to me yes.
So, the behavior is what you describe:
when "private" is set, we don't care anymore about what "public" is (and we shouldn't, that'd be too complex if we did, that's what I meant with "higher precedence".)
when "private" is set, the service is not inlined nor removed anymore (but that's internal details we shouldn't tell about), and as a consequence, the service is private but remains publicly accessible, with a deprecation notice along the way, which is exactly what we want.

nicolas-grekas replied:
reworded. better?

nicolas-grekas commented:
PR is now ready and tested.
Please help validate the list of still public services/aliases (see PR)

lyrixx commented on REMAINING.PUBLIC.md:
```diff
@@ -0,0 +1,60 @@
+security.password_encoder
+session.handler
+session.storage
+templating.loader
```
IMHO, all theses services (above) could be private. except the `security.password_encoder`  (for use in ContainerAware classes)

nicolas-grekas commented:
All services are now either private, or public with a justification notice (see PR.)
Calling @symfony/deciders for review+votes before I remove that list of public services from the patch.

nicolas-grekas commented:
failures are false positives
list of remaining public services added to the PR's description and removed from patch
ready for vote+merge

javiereguiluz commented:
Very nice work! Imagining myself as someone new to Symfony, I would have issues/doubts with these services:

1) Everything related to "cache clear": it looks like something internal to Symfony which I shouldn't care about

`cache.app_clearer` alias
`cache_clearer` <- for the cache:clear command
`cache.global_clearer` <- for the cache:pool:clear command to clear all pools

2) Controllers: all these services

`twig.controller.*`
`web_profiler.controller.*`

chalasr commented:
@javiereguiluz We still have some places where we are forced to `get()` public services from the container internally, mostly places where we fetch unpredictable services which cannot be grouped by scope, like controllers which can be any service (no required tag), similar problem for those related to cache (have a look at CacheClearCommand/CachePoolClearCommand).
The services you listed must remain public.

ro0NL commented:
command related services can be made private in 4.0 :)

mvrhov commented:
If the controllers are the only big area which cannot be made private.. Maybe the controller tag should be introduced to make this possible?

nicolas-grekas commented:
There is already a tag to flag services as controllers (`controller.service_arguments`).
BUT it works the other way around: the **routing** contains references to services.
So that the correct logic to do that would be to parse the routing config to flag controllers as such.
That's not something we have right now, and I'm personally no planning to work on it at all: controllers as public services are just fine, there is nothing to "fix" here.

mvrhov commented:
Just to complete this. But `controller.service_arguments` doesn't work for the controllers that are actually registered as services. And I find it weird that I'd have to tag with that when using "actions" for controllers e.g one class per action with `__invoke` and all other dependencies injected via constructor.

nicolas-grekas commented:
@mvrhov I don't know anymore what you're talking about :) You don't have to use that tag at all, that's what   I said. You just have to make your controllers public.

javiereguiluz commented:
@nicolas-grekas the thing that controllers must be public is more confusing than it looks. I can understand public services: you may inject this somewhere, so make it public. OK. But I don't inject or need or use or do anything with controllers. It's "a Symfony thing" and that's why I expect them to not be public.

nicolas-grekas commented:
OK, I've no solution for that - and I don't think there is anything wrong in fact.

A service really needs to public for one and only one reason: bootstrapping.
That's because the app must start with something.
There are at least three common bootstrapping contexts:
- HTTP : here the entry point is routing thus controllers
- CLI: the entry point is the Application thus commands
- tests: the entry point is whatever you use in the body of your test cases

all other uses are the bad sort of service locator.

nicolas-grekas commented:
ping @symfony/deciders, one vote missing :)

kbond commented:
@nicolas-grekas After upgrading to 3.4, I'm getting this deprecation warning:

```
User Deprecated: The "routing.loader" service is private, getting it from the container is deprecated since Symfony 3.2 and will fail in 4.0. You should either make the service public, or stop using the container directly and use dependency injection instead.
```

This is the culprit:

https://github.com/symfony/symfony/blob/78426b79d2a66c2785bce66ed18bf869b3800ec4/src/Symfony/Bundle/FrameworkBundle/Routing/Router.php#L56

Tobion commented:
The `routing.loader` should be public. So that is strange.

kbond commented:
Yeah, I discovered a bundle is at fault... Please ignore...
